tests/src/run.rs

```diff
@@ -125,33 +138,131 @@ impl<'a> Runner<'a> {
     fn handle_not_emitted(&mut self) {
         let mut first = true;
         for (note, &seen) in self.test.notes.iter().zip(&self.seen) {
-            if seen {
+            let mut missing_stages = TestStages::empty();
+            let required_stages = ARGS.stages() & self.test.attrs.stages;
+
+            // If a test stage requiring the paged target has been specified,
+            // require the annotated diagnostics to be present in any paged
+            // stage. For example if `pdftags` is specified and an error is
+            // emitted in the pdf stage, that is ok.
+            if !required_stages.paged_stages().is_empty()
+                && (TestStages::all().paged_stages() & seen).is_empty()
+            {
+                missing_stages |= TestStages::PAGED;
+            }
+            if !required_stages.html_stages().is_empty()
+                && (TestStages::all().html_stages() & seen).is_empty()
+            {
+                missing_stages |= TestStages::HTML;
+            }
+
+            if missing_stages.is_empty() {
                 continue;
             }
             let note_range = self.format_range(note.file, &note.range);
             if first {
-                log!(self, "not emitted");
+                log!(self, "not emitted [{missing_stages}]");
                 first = false;
             }
             log!(self, "  {}: {note_range} {} ({})", note.kind, note.message, note.pos,);
         }
     }
 
+    fn compile<D: Document>(&mut self, target: TestTarget) -> Option<D> {
+        let Warned { output, warnings } = typst::compile::<D>(&self.world);
+        for warning in &warnings {
+            self.check_diagnostic(NoteKind::Warning, warning, target);
+        }
+
+        if let Err(errors) = &output {
+            for error in errors.iter() {
+                self.check_diagnostic(NoteKind::Error, error, target);
+            }
+        }
+
+        output.ok()
+    }
+
+    fn run_file_test<T: FileOutputType>(
+        &mut self,
+        doc: &Option<T::Doc>,
+    ) -> Option<T::Live> {
+        let output = self.run_test::<T>(doc);
+        if self.test.attrs.should_check_ref(T::OUTPUT) {
+            self.check_file_ref::<T>(&output)
+        }
+        output.map(|(_, live)| live)
+    }
+
+    fn run_hash_test<T: HashOutputType>(
+        &mut self,
+        doc: &Option<T::Doc>,
+    ) -> Option<T::Live> {
+        let output = self.run_test::<T>(doc);
+        if self.test.attrs.should_check_ref(T::OUTPUT) {
+            self.check_hash_ref::<T>(&output)
+        }
+        output.map(|(_, live)| live)
+    }
+
+    /// Run test specific to document format.
+    fn run_test<T: OutputType>(
+        &mut self,
+        doc: &Option<T::Doc>,
+    ) -> Option<(T::Doc, T::Live)> {
+        let live_path = T::OUTPUT.live_path(&self.test.name);
+
+        let output =
+            doc.clone()
```

Comment:
if we clone anyway, might as well take an owned doc. will also format more nicely.